src/main/java/org/opensearch/security/auth/SecurityUser.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ */
+package org.opensearch.security.auth;
+
+import java.security.Principal;
+import java.util.concurrent.Callable;
+
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.identity.NamedPrincipal;
+import org.opensearch.identity.UserSubject;
+import org.opensearch.identity.tokens.AuthToken;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.user.User;
+import org.opensearch.threadpool.ThreadPool;
+
+public class SecurityUser implements UserSubject {
+    private final NamedPrincipal userPrincipal;
+    private final ThreadPool threadPool;
+    private final User user;
+
+    SecurityUser(ThreadPool threadPool, User user) {
+        this.threadPool = threadPool;
+        this.user = user;
+        this.userPrincipal = new NamedPrincipal(user.getName());
+    }
+
+    @Override
+    public void authenticate(AuthToken authToken) {
+        // not implemented
+    }
+
+    @Override
+    public Principal getPrincipal() {
+        return userPrincipal;
+    }
+
+    @Override
+    public <T> T runAs(Callable<T> callable) throws Exception {
```

Comment:
I am wondering if this `runAs()` API covers all possible scenarios. 

It is sufficient for synchronous scenarios, where it allows to run the code subsequent to this `runAs()` call again in the original context. 

However, OpenSearch architecture is fundamentally asynchronos. Often, it happens that "subsequent" logic does actually not follow in the subsequent lines but in the `onResult()` action listener method. In such a `onResult()` method, there would be no way to restore the context with this `runAs()` API.

Using the `stashContext()` API dirtectly however supports such scenarios, one example of this pattern is here (see the restore call in line 238):

https://github.com/opensearch-project/security/blob/2fa7effd0dacc7cb76e0967579834865b4a09d63/src/main/java/org/opensearch/security/configuration/DlsFilterLevelActionHandler.java#L166-L257

Note: I understand that this example does not match the use case for `runAs()` - however I can imagine that also plugins could need such scenarios. Or do I get something wrong?

Reply:
To illustrate the issue, look at the following code (from the int test):

https://github.com/opensearch-project/security/blob/154556c091afd5eff1569ce787ae4c0ccdeba96e/src/integrationTest/java/org/opensearch/security/plugin/TransportRunClusterHealthAction.java#L51-L84

If `runAs` equals `plugin`, then the `onResponse()` method of the provided `actionListener` would be executed in the plugin user context. Thus, any client which calls this action would suddenly have a plugin user scope.

Reply:
The ultimate goal is to ensure that authz checks are run with any transport request that is executed. The security plugin needs to have access to an identity associated with the executor of a transport action. Currently with stashContext, the identity is erased form the TC and any plugin can operate w/o authz checks.

What is the purpose of restoring the context in DlsFilterLevelActionHandler? Is restoring the context necessary? wdyt about a higher-level api (similar to runAs) where it can be guaranteed when restoring the context that the OPENDISTRO_SECURITY_USER transient header is populated to ensure that authz checks are performed?

Reply:
> What is the purpose of restoring the context in DlsFilterLevelActionHandler? Is restoring the context necessary? 

Yes, because the control is returned then to the caller of the action. The caller might call another action for which DLS/FLS also needs to be enforced. This requires the `ConfigConstants.OPENDISTRO_SECURITY_FILTER_LEVEL_DLS_DONE` header to be absent. If it would be still there, no DLS/FLS protections would be applied.

Just, to avoid any confusions: I used this example only to illustrate the pattern - a pattern which is IMHO an essential pattern.

The `runAs()` API has IMHO the fundamental issue that it assumes a single threading model. However, single thread scenarios are rather the exception than the norm in OpenSearch. IMHO, there needs to be another way to tell "this block of logic has finished" other than the current thread leaving the `runAs()` block again.

Reply:
That's a good point. I just tested out the restore approach you posted above and it works within a `runAs` block.

i.e.

```
...
} else if ("plugin".equalsIgnoreCase(runAs)) {
    ThreadContext.StoredContext ctx = threadPool.getThreadContext().newStoredContext(true);
    ActionListener<ClusterHealthResponse> chr = ActionListener.wrap(
            r -> {
                ctx.restore();
                actionListener.onResponse(new RunClusterHealthResponse(true)); },
            e -> {
                System.out.println("Before restore");
                System.out.println("Current user: " + threadPool.getThreadContext().getTransient(OPENDISTRO_SECURITY_USER));
                ctx.restore();
                System.out.println("After restore");
                System.out.println("Current user: " + threadPool.getThreadContext().getTransient(OPENDISTRO_SECURITY_USER));
                actionListener.onFailure(e);
            }
    );
    contextSwitcher.runAs(() -> {
        client.admin().cluster().health(new ClusterHealthRequest(), chr);
        return null;
    });
}
...
```

and in the output I see

```
> Task :integrationTest

SystemIndexTests > testPluginShouldNotBeAbleToRunClusterActions STANDARD_OUT
    Before restore
    Current user: User [name=plugin:org.opensearch.security.systemindex.sampleplugin.SystemIndexPlugin1, backend_roles=[], requestedTenant=null]
    After restore
    Current user: User [name=admin, backend_roles=[], requestedTenant=null]
```

Ideally, there is a higher-level API when executing a transport action to ensure the handler for the transport response is run in the original context, but the same workaround from your origin comment works here as well. This PR would incrementally improve the security posture by ensuring that the transport action is run with authz checks and ensuring that a plugin only accesses its own system indices.